### ISSUE TITLE
[DRY] Refactor algebraic properties of types

### ISSUE THREAD

**Issue by jamesmckinna** [agda/agda-stdlib#2489]
There's (a lot of) duplication between `Function.Related.TypeIsomorphisms` and `Data.Product.Algebra` (and possibly elsewhere: eg the duplication of `True↔` between the first of these, and `Relation.Nullary.Decidable`...).

This might be bad enough to want to deprecate whole modules, but there is a subtle problem exposed by #2419 , namely that not all the lemmas are as level-polymorphic as they might be, esp. those in `Data.Product.Algebra`, where the use of `Algebra.Definitions` imposes a homogeneous level restriction.

So a refined refactoring might be to make`Function.Related.TypeIsomorphisms` home for the heterogeneous level-polymorphic definitions (the direction in #2419 ), while making `Data.Product.Algebra` home for the homogeneous/algebraic versions?

This feels as though it might be a useful clean-up, but potentially `breaking` wrt possibly unresolved `Level` metas and/or name clashing which no obvious (or at least: sensible) deprecation strategy can comfortably accommodate...

**Comment by JacquesCarette** [comment]
`Function.Related.TypeIsomorphism` is both quite old and very useful. The level polymorphism is definitely something worth having.

I wonder if the (deeper) issue is that type "algebra" isn't quite 1-sorted algebra. So that our current `Data.*.Algebra` is correct wrt the usual notion of 'algebra', but we're encountering a genuine generalization here.

**Comment by jamesmckinna** [comment]
> `Function.Related.TypeIsomorphism` is both quite old and very useful. The level polymorphism is definitely something worth having.

I agree (re: level polymorphism... but where these things live is moot, AFAIAC)

**Comment by jamesmckinna** [comment]
> I wonder if the (deeper) issue is that type "algebra" isn't quite 1-sorted algebra. So that our current `Data.*.Algebra` is correct wrt the usual notion of 'algebra', but we're encountering a genuine generalization here.

I'm not sure I quite understand this: some things (esp. `Level`-polymorphic such) are 'large' notions (moreover: typically 'type constructors' in some form or another), whereas relative to a given carrier, the same things are 'small', and more like 'properties'/'structure'...? I was trying to cash that distinction out by saying we should... split them between different modules!?

**Comment by JacquesCarette** [comment]
If Agda had `--cumulativity` as default (which it doesn't, and I'm fine with that), one could see the stuff in `Function.Related.TypeIsomorphism` as 'algebra' for `Set\omega`.

In any case, it seems we agree: even though there is duplication, it should exist. Just elsewhere.

**Comment by jamesmckinna** [comment]
It seems that #2868 partially tackles this, taking an alternative approach, but one certainly in the spirit of this issue: make the `Data.*` proofs more `Level`-polymorphic, and then re-export them.

**Comment by JacquesCarette** [comment]
Given that #2868 has been merged, what would be left to do before closing this issue?

**Comment by jamesmckinna** [comment]
I'm not entirely sure! (too much to page back in after too long, and the PR went in the opposite direction to my thinking on the subject...).

I suppose that there might still be an outstanding question as to whether the algebra of `Sum` and `Product` together are enough make `Function.Related.TypeIsomorphism` redundant/obsolete/deprecatable. But let's close and return to such an issue afresh, if necessary?
